lib/rules/require-key-for-conditionally-rendered-repeated-components.js

```diff
@@ -0,0 +1,371 @@
+/**
+ * @author Felipe Melendez
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+// =============================================================================
+// Requirements
+// =============================================================================
+
+const utils = require('../utils')
+const casing = require('../utils/casing')
+const reservedNamesInVue2 = require('../utils/vue2-builtin-components')
+
+// =============================================================================
+// Constants and Variables
+// =============================================================================
+
+/**
+ * Curated list of standard HTML elements that serve as containers.
+ * We're not using an existing library util because we are only concerned about HTML elements that
+ * have the capacity to wrap other components.
```

Comment:
Why? Shouldn't this also require a key with this rule?

```html
<span v-if="foo">Foo <strong>foo</strong></span>
<span v-else>Bar <strong>bar</strong></span>
```

Reply:
Hi! Thank you for taking the time to review my PR. I appreciate your attention to detail.

The key attribute becomes crucial when we deal with components or elements that hold internal state or have side-effects. Providing a unique key for each condition ensures that Vue treats each instance separately, preventing unintended state persistence or side-effects between condition changes. However, in straightforward cases like span elements, which don't have internal state or side-effects, the key attribute is not necessary for correct rendering behavior.

Reply:
But shouldn't we restrict the rule to actual components then and not bother with `div`s etc. at all?

Reply:
yep, great point, the rule does not target standard HTML tags. It specifically targets custom Vue components that are used within conditional rendering contexts.
The rule defines a list of native HTML tags and reserved Vue component names as container elements, and there is a function `isContainerElement` that checks if a given node (representing a Vue component or element) is a container element. If the current node represents a container element, the function returns early, effectively skipping any further processing for that node. This ensures that the rule's logic for checking and enforcing the key attribute is only applied to custom Vue components, not native HTML tags or reserved Vue names.
I created a [test suite for this rule](https://gist.github.com/felipemelendez/981d58a4ae38782323b0df5d820ecfea), perhaps that will bring further clarity as to what is expected of the rule.

Reply:
ahh, oops, good catch, I created this rule for our app at work and it appears that I changed it so that it targets divs, it's been a while, and I can't remember why I did that but you are absolutely correct, my bad, I will fix it as soon as I get a chance. Thank you for making that observation

Reply:
ahh, that is a great observation, good catch, I will make that update, and refactor tests accordingly

Reply:
[0daeb91](https://github.com/vuejs/eslint-plugin-vue/pull/2280/commits/0daeb91f54d1781a733f3de87f615e3f575b2cd0)